File: 1011 Capacity To Ship Packages Within D Days/1011capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int ans = INT_MAX;

    int check(vector<int> arr,int c){
        int d = 0;
        int cur = 0;
        for(int i=0;i<arr.size();i++){
            if(c<arr[i])
                return -1;
            cur+=arr[i];
            if(cur==c){
                d+=1;
                cur=0;
            }
            else if(cur>c){
                d+=1;
                cur = arr[i];
            }   
        }
        if(cur>0 && cur<=c)
            d+=1;
        else if(cur>c)
            return -1;
        return d;
    }
    
    void bin(int s,int e,int h,vector<int> arr,int c,int mc){
        while(s<=e){
            int mid = s+(e-s)/2;
            if(mid<mc){
                s=mid+1;
                continue;
            }  
            int val = check(arr,mid);
            if(val==-1 || val>h)
                s=mid+1;
            else if(val<=h){
                ans = min(ans,mid);
                e = mid-1;
            }
        }
    }

    int shipWithinDays(vector<int>& weights, int days) {
        int cap = 0;
        int mc = INT_MIN;
        for(int i=0;i<weights.size();i++){
            cap+=weights[i];
            mc = max(mc,weights[i]);
        }
        int n = weights.size();
        int s = 1;
        int e = cap;
        bin(s,e,days,weights,cap,mc);
        return ans;
    }
};

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        '''
        binary search on weight capacity of the ship

        l = max(weights)
        right = sum(weights)

        need function possible - 
        it takes weight capacity
        for each weight capacity it calculates days taken to ship all the wights
        if days_taken <= days return True, else False

        when possible is True - right = mid
        else left = mid + 1
        '''

        def possible(capacity):
            days_taken = 0
            cur_capacity = capacity
            for weight in weights:
                if weight > cur_capacity:
                    cur_capacity = capacity
                    days_taken += 1
                
                cur_capacity -= weight
            
            days_taken += 1
            print(capacity, days_taken)
            return days_taken <= days
        
        left = max(weights)
        right = sum(weights)

        while left < right:
            mid = left + (right - left)//2

            if possible(mid):
                right = mid
            else:
                left = mid + 1
        
        return left
